src/delphi/eval/utils.py

```diff
@@ -42,3 +54,42 @@ def load_validation_dataset(dataset_name: str) -> Dataset:
         split="train",
     )
     return cast(Dataset, dataset)
+
+
+def load_text_from_dataset(dataset: Dataset) -> list[str]:
+    text = []
+    for sample_txt in dataset["story"]:
+        # encoding issues and rare weird prompts
+        if not set(sample_txt).issubset(ALLOWED_CHARS):
+            continue
+        text.append(sample_txt)
+    return text
+
+
+def tokenize(
+    tokenizer: PreTrainedTokenizerBase, sample_txt: str
+) -> Int[torch.Tensor, "seq"]:
+    # supposedly this can be different than prepending the bos token id
+    return cast(
+        Int[torch.Tensor, "seq"],
+        tokenizer.encode(tokenizer.bos_token + sample_txt, return_tensors="pt")[0],
+    )
+
+
+def get_correct_and_all_probs(
+    model: PreTrainedModel, sample_tok: Int[torch.Tensor, "seq"]
+) -> tuple[Float[torch.Tensor, "next_seq"], Float[torch.Tensor, "next_seq vocab"]]:
+    """Get probabilities for the actual next token and for all predictions"""
+    # remove the first token (the bos token)
+    probs = get_single_logprobs(model, sample_tok)[1:]
+    correct_probs = probs[range(len(probs)), sample_tok[1:]]
+    return correct_probs, probs
+
+
+def get_correct_and_top_probs(
+    model: PreTrainedModel, sample_tok: Int[torch.Tensor, "seq"], top_k: int = 3
+) -> tuple[Float[torch.Tensor, "next_seq"], torch.return_types.topk]:
+    """Get probabilities for the actual next token and for top k predictions"""
+    correct_probs, probs = get_correct_and_all_probs(model, sample_tok)
+    top_k_probs = torch.topk(probs, top_k, dim=-1)
+    return correct_probs, top_k_probs
```

Comment:
too many functions
- you don't need get_single_logprobs
- you can rename get_next_logprobs to get_all_and_next_logprobs and make it return both
- make a wrapper get_next_logprobs around get_all_and_next_logprobs that drops the first return value
- then get_next_and_top_probs can call get_all_and_next_logprobs, do squeeze/unsqueeze and apply exp() to get probs from logprobs

Reply:
There's a surprising problem with this that I've just spent entirely too much time chasing down: the `topk` object is constructed in the topk function call and is otherwise seemingly impossible to create. The implementation is getting into cpython magic deep in torch, but basically we can't create a new topk object outside of running topk. That means we need to either re-run the topk call, or wreck all of the subsequent topk type hints.

I've tried everything I can think of to get a topk object out of this and spent entirely too much time on it. So: I'm just adding a `remove_batch_dim` flag to `get_next_and_top_k_probs`.

Reply:
Ugh, it's actually worse than this. The topk object is run on the logprobs, which is what then ends up in the visualization with no way to change it. So `get_next_and_top_probs` is going to be slightly redundant and just not wrap `get_all_and_next_logprobs` at all.